### PR TITLE
modules: add torch.nn.Linear interface

### PR DESCRIPTION
**Pull request by amo33**
This PR adds torch.nn.Linear layer. 

After searching megatron-LM and other repos it has been obvious that unpadding and padding process is inefficient. Also, just using python package flash-attn seems adequate.

Below is an additional experiment.
Variable length inputs can be handled by one fused cuda kernel in Flash-Attention.
As FlatFlow doesn't need to stack data samples, we design to exclude unpadding and padding. 
Unit test is conducting comparison between Flash-Attention's interface and FlatFlow's interface.
Comparison is about cuda memory usage and cuda latency.
This unit test is conducted under Nvidia A100 environment.
Output using ```pytest tests``` :
```
===================================================================================================================== test session starts =====================================================================================================================
platform linux -- Python 3.9.19, pytest-8.2.2, pluggy-1.5.0
rootdir: /home/flatflow/tests
configfile: pytest.ini
collected 768 items                                                                                                                                                                                                                                           

tests/test_flash_attn.py::test_flash_attn_varlen_output[0.0-108-216-32-False-True-True-False-mha-dtype0-False] PASSED       [  0%]
tests/test_flash_attn.py::test_flash_attn_varlen_output[0.0-108-216-32-False-True-True-False-mqa-dtype0-False] PASSED       [  0%]
tests/test_flash_attn.py::test_flash_attn_varlen_output[0.0-108-216-32-False-True-True-False-gqa-dtype0-False] PASSED       [  0%]                                                                                                              
...

tests/test_flash_attn.py::test_flash_attn_varlen_output[0.17-256-512-256-True-True-True-True-mqa-dtype0-False] PASSED       [ 99%]
tests/test_flash_attn.py::test_flash_attn_varlen_output[0.17-256-512-256-True-True-True-True-gqa-dtype0-False] PASSED       [100%]
Average CUDA time improvement: 93.03%
Average CUDA memory improvement: 80.07%


==================================================================================================================== 768 passed in 36.14s =====================================================================================================================
```
As above log by skipping padding and unpadding process, latency and memory improvement is shown.

### CODE REVIEW

**Review by 9rum — COMMENTED** [review]
@amo33 Thanks for filing this PR.
For now, it would be more beneficial to review the overall structure rather than reviewing individual line of code.

Regarding the ``flatflow/torch/nn`` package, we aim to provide compatible interfaces with the existing PyTorch implementation, except for providing flattened tensor and offsets together. Therefore, ``flatflow/torch/nn/custom_linear.py`` should be moved to ``flatflow/torch/nn/modules/linear.py`` and should be same as PyTorch's ``nn.Linear`` interface (except for the offsets). The same goes for FlashAttention modules.

The ``tests`` package is more of benchmarks than actual tests. Since we are still in the early stage of implementation, it would be better to keep the benchmark scripts in your local for now. Dependency management is not necessary at this stage, so I recommend to exclude the testing directory and ``requirements.txt`` from this PR.

**Comment by amo33** [comment]
@9rum Thanks for your detailed review! 
Current PR is updated based on your review.

**Review by 9rum — COMMENTED** [review]
@amo33 It seems that the arguments and return types of ``Linear.forward`` need to be changed.
Our programming model is designed to provide ``offsets``, which are ``List[int]``, so the arguments of ``forward`` must receive ``input: Tensor`` and ``offsets: List[int]``, and return a ``tuple[Tensor, List[int]]``.

**Comment by amo33** [comment]
@9rum Thanks for the comment.
I was expecting at least individual List[int]. 
I agree for excluding max_seqlen_q and max_seqlen_k. 
On the other hand for cu_seqlens_q and cu_seqlens_k  we need to consider gqa , so shouldn't these two need to be separately stored?

**Review by 9rum — COMMENTED** [review]
LGTM
